### PR TITLE
feat(metrics): Add caching to project transaction thresholds [TET-515]

### PR DESCRIPTION
**Pull request by iambriccardo**
This PR aims at fixing a performance regression problem that resulted into multiple queries executed during `apdex` snql query generation. The set of duplicate queries is shown on [Sentry](https://sentry.io/organizations/sentry/discover/sentry:8dc11eb1ecc3439085f3f12dacdea6c4/?field=transaction&field=timestamp&homepage=true&id=16351&interval=15m&name=BlahBlah&project=1&query=event.type:transaction+user.email:william.mak@sentry.io+transaction:/api/0/organizations/%7Borganization_slug%7D/events/&sort=-timestamp&statsPeriod=24h&topEvents=5&yAxis=count()#span-83da759280f1fde5).

In order to solve this issue, caching has been implemented, with a TTL of 1 hour, that was set under the assumption that project thresholds do not change very frequently and also that having an out-of-date copy of them for at maximum 1 hour is not a problem. Caching is implemented via the in-memory cache and a copy of the data is saved into the cache when it is not present, where presence is determined by a cache key that is uniquely generated by combining all the input parameters of the function (in order).

### CODE REVIEW

File: src/sentry/models/transaction_threshold.py [hunk 1]
```diff
@@ -15,6 +17,37 @@ class TransactionMetric(Enum):
...
+PROJECT_TRANSACTION_THRESHOLD_CACHE_TTL = 3600
+
+
+def _get_cache_key(model_name, project_ids, org_id, order_by, value_list):
+    return "{}:{}".format(
+        model_name, md5_text(f"{project_ids}:{org_id}:{order_by}:{value_list}").hexdigest()
+    )
+
+
+def _filter_and_cache(cls, cache_key, project_id__in, organization_id, order_by, value_list):
```
Comment (ahmedetefy) [inline src/sentry/models/transaction_threshold.py:34]: Please remove all naming to `project_id__in` everywhere to just `project_ids` except for when you are making this query

```suggestion
def _filter_and_cache(cls, cache_key, project_ids, organization_id, order_by, value_list):
```

File: tests/sentry/snuba/metrics/test_snql.py [hunk 2]
```diff
@@ -434,6 +437,41 @@ def test_dist_count_aggregation_on_tx_satisfaction(self):
             "transaction.tolerated",
         )
 
+    @patch("sentry.models.transaction_threshold.ProjectTransactionThresholdOverride.objects.filter")
```
Comment (ahmedetefy) [inline tests/sentry/snuba/metrics/test_snql.py:440]: There is a missing test here, can you please create a test that fetches the project transaction threshold once and then examine the contents of the cache and then ensure that the next call gets the value from the cache

**Review by ahmedetefy — CHANGES_REQUESTED** [review]
Logic seems okay in general. Please add a better description

File: src/sentry/models/transaction_threshold.py (hunk 1 above)
Comment (iambriccardo) [reply src/sentry/models/transaction_threshold.py:34]: I agree! Thanks

File: tests/sentry/snuba/metrics/test_snql.py (hunk 2 above)
Comment (iambriccardo) [reply tests/sentry/snuba/metrics/test_snql.py:440]: Yep! Thanks

**Review by ahmedetefy — APPROVED** [review]

**Pull request merged**